game/graphics/texture/TexturePool.cpp

```diff
@@ -0,0 +1,195 @@
+    return true;
+  }
+};
+
+/*!
+ * Handle a GOAL texture-page object being uploaded to VRAM.
+ * The strategy:
+ * - upload the texture-age to a fake 4MB VRAM, like the GOAL code would have done.
+ * - "download" each texture in a reasonable format for the PC Port (currently RGBA8888)
+ * - add this to the PC pool.
+ *
+ * The textures are scrambled around in a confusing way.
+ *
+ * NOTE: the actual conversion is currently done here, but this might be too slow.
+ * We could store textures in the right format to begin with, or spread the conversion out over
+ * multiple frames.
+ */
+void TexturePool::handle_upload_now(const u8* tpage, int mode, const u8* memory_base, u32 s7_ptr) {
+  Timer timer;
+
+  // extract the texture-page object. This is just a description of the page data.
+  GoalTexturePage texture_page;
+  memcpy(&texture_page, tpage, sizeof(GoalTexturePage));
+
+  u32 sizes[3] = {texture_page.segment[0].size, texture_page.segment[1].size,
+                  texture_page.segment[2].size};
+  if (mode == -1) {
+    // I don't really understand what's going on here with the size.
+    // the sizes given aren't the actual sizes in memory, so if you just use that, you get the
+    // wrong answer. I solved this in the decompiler by using the size of the actual data, but we
+    // don't really have that here.
+    u32 size = ((sizes[0] + sizes[1] + sizes[2] + 255) / 256) * 256;
+    m_tex_converter.upload(memory_base + texture_page.segment[0].block_data_ptr,
+                           texture_page.segment[0].dest, size);
+  } else {
+    // no reason to skip this, other than
+    lg::error("TexturePool skipping upload now with mode {}.", mode);
+    return;
+  }
+
+  // loop over all texture in the tpage and download them.
+  for (int tex_idx = 0; tex_idx < texture_page.length; tex_idx++) {
+    GoalTexture tex;
+    if (texture_page.try_copy_texture_description(&tex, tex_idx, memory_base, tpage, s7_ptr)) {
+      // each texture may have multiple mip levels.
+      for (int mip_idx = 0; mip_idx < tex.num_mips; mip_idx++) {
+        u32 ww = tex.w >> mip_idx;
+        u32 hh = tex.h >> mip_idx;
+        u32 size_bytes = ww * hh * 4;
+
+        auto texture_record = std::make_unique<TextureRecord>();
+        texture_record->page_name = goal_string(texture_page.name_ptr, memory_base);
+        texture_record->name = goal_string(tex.name_ptr, memory_base);
+        texture_record->mip_level = mip_idx;
+        texture_record->w = ww;
+        texture_record->h = hh;
+        texture_record->data_segment = tex.segment_of_mip(mip_idx);
+        texture_record->data.resize(size_bytes);
+
+        m_tex_converter.download_rgba8888(texture_record->data.data(), tex.dest[mip_idx],
+                                          tex.width[mip_idx], ww, hh, tex.psm, tex.clutpsm,
+                                          tex.clut_dest, size_bytes);
+        set_texture(tex.dest[mip_idx], std::move(texture_record));
```

Comment:
`texture_record` is used after this line for the debug texture dumping, and this seems to be causing some sort of undefined behavior (even if debug output is disabled).

Reply:
Good catch this is definitely wrong.